File: CS/ThemeSelector.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;

namespace Styling {
    public class ThemeResourceDictionary : ResourceDictionary { }

    public class ThemeSelector : DependencyObject {
        public static readonly DependencyProperty CurrentThemeDictionaryProperty = DependencyProperty.RegisterAttached(
            "CurrentThemeDictionary",
            typeof(Uri),
            typeof(ThemeSelector),
            new UIPropertyMetadata(null, CurrentThemeDictionaryChanged));

        public static Uri GetCurrentThemeDictionary(DependencyObject obj) {
            return (Uri)obj.GetValue(CurrentThemeDictionaryProperty);
        }

        public static void SetCurrentThemeDictionary(DependencyObject obj, Uri value) {
            obj.SetValue(CurrentThemeDictionaryProperty, value);
        }

        private static void CurrentThemeDictionaryChanged(DependencyObject obj, DependencyPropertyChangedEventArgs e) {
            if (obj is FrameworkElement) {
                ApplyTheme(obj as FrameworkElement, GetCurrentThemeDictionary(obj));
            }
        }

        private static void ApplyTheme(FrameworkElement targetElement, Uri dictionaryUri) {
            ThemeResourceDictionary themeDictionary = new ThemeResourceDictionary();

            if (dictionaryUri != null) {
                themeDictionary.Source = dictionaryUri;

                // add the new dictionary to the collection of merged dictionaries of the target object
                targetElement.Resources.MergedDictionaries.Insert(0, themeDictionary);
            }

            // find if the target element already has a theme applied
            List<ThemeResourceDictionary> existingDictionaries =
                (from dictionary in targetElement.Resources.MergedDictionaries.OfType<ThemeResourceDictionary>()
                 select dictionary).ToList();

            // remove the existing dictionaries 
            foreach (ThemeResourceDictionary thDictionary in existingDictionaries) {
                if (themeDictionary == thDictionary) continue;  // don't remove the newly added dictionary
                targetElement.Resources.MergedDictionaries.Remove(thDictionary);
            }
        }
    }
}

File: VB/ThemeSelector.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows

Namespace Styling

    Public Class ThemeResourceDictionary
        Inherits ResourceDictionary

    End Class

    Public Class ThemeSelector
        Inherits DependencyObject

        Public Shared ReadOnly CurrentThemeDictionaryProperty As DependencyProperty = DependencyProperty.RegisterAttached("CurrentThemeDictionary", GetType(Uri), GetType(ThemeSelector), New UIPropertyMetadata(Nothing, AddressOf CurrentThemeDictionaryChanged))

        Public Shared Function GetCurrentThemeDictionary(ByVal obj As DependencyObject) As Uri
            Return CType(obj.GetValue(CurrentThemeDictionaryProperty), Uri)
        End Function

        Public Shared Sub SetCurrentThemeDictionary(ByVal obj As DependencyObject, ByVal value As Uri)
            obj.SetValue(CurrentThemeDictionaryProperty, value)
        End Sub

        Private Shared Sub CurrentThemeDictionaryChanged(ByVal obj As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
            If TypeOf obj Is FrameworkElement Then
                Call ApplyTheme(TryCast(obj, FrameworkElement), GetCurrentThemeDictionary(obj))
            End If
        End Sub

        Private Shared Sub ApplyTheme(ByVal targetElement As FrameworkElement, ByVal dictionaryUri As Uri)
            Dim themeDictionary As ThemeResourceDictionary = New ThemeResourceDictionary()
            If dictionaryUri IsNot Nothing Then
                themeDictionary.Source = dictionaryUri
                ' add the new dictionary to the collection of merged dictionaries of the target object
                targetElement.Resources.MergedDictionaries.Insert(0, themeDictionary)
            End If

            ' find if the target element already has a theme applied
            Dim existingDictionaries As List(Of ThemeResourceDictionary) =(From dictionary In targetElement.Resources.MergedDictionaries.OfType(Of ThemeResourceDictionary)() Select dictionary).ToList()
            ' remove the existing dictionaries 
            For Each thDictionary As ThemeResourceDictionary In existingDictionaries
                If themeDictionary Is thDictionary Then Continue For ' don't remove the newly added dictionary
                targetElement.Resources.MergedDictionaries.Remove(thDictionary)
            Next
        End Sub
    End Class
End Namespace
